File: cdp-agentkit-core/typescript/src/actions/cdp/deploy_contract.ts
import { CdpAction } from "./cdp_action";
import { Wallet } from "@coinbase/coinbase-sdk";
import { z } from "zod";

const DEPLOY_CONTRACT_PROMPT = `
Deploys smart contract with required args: solidity version (string), solidity input json (string), contract name (string), and optional constructor args (Record<string,any>).

Input json structure:
{"language":"Solidity","settings":{"remappings":[],"outputSelection":{"*":{"*":["abi","evm.bytecode"]}}},"sources":{}}

The solidity version must be >= 0.8.0 and <= 0.8.28.

You must set the abi and evm.bytecode in the outputSelection. Do not include any extra spaces in the JSON key fields. Remappings can be set as needed in settings. 
Sources should contain inline contract code. Include library source if needed. If the constructor takes in parameters, you must pass in the constructor args, which is a key-value
map where the key is the argument name and the value is the argument value. Encode uint/int/bytes/string/address values as strings, boolean values as true/false. For arrays/tuples, encode based on contained type.
`;

const SolidityVersions = {
  "0.8.28": "0.8.28+commit.7893614a",
  "0.8.27": "0.8.27+commit.40a35a09", 
  "0.8.26": "0.8.26+commit.8a97fa7a",
  "0.8.25": "0.8.25+commit.b61c2a91",
  "0.8.24": "0.8.24+commit.e11b9ed9",
  "0.8.23": "0.8.23+commit.f704f362",
  "0.8.22": "0.8.22+commit.4fc1097e",
  "0.8.21": "0.8.21+commit.d9974bed",
  "0.8.20": "0.8.20+commit.a1b79de6",
  "0.8.19": "0.8.19+commit.7dd6d404",
  "0.8.18": "0.8.18+commit.87f61d96",
  "0.8.17": "0.8.17+commit.8df45f5f",
  "0.8.16": "0.8.16+commit.07a7930e",
  "0.8.15": "0.8.15+commit.e14f2714",
  "0.8.14": "0.8.14+commit.80d49f37",
  "0.8.13": "0.8.13+commit.abaa5c0e",
  "0.8.12": "0.8.12+commit.f00d7308",
  "0.8.11": "0.8.11+commit.d7f03943",
  "0.8.10": "0.8.10+commit.fc410830",
  "0.8.9": "0.8.9+commit.e5eed63a",
  "0.8.8": "0.8.8+commit.dddeac2f", 
  "0.8.7": "0.8.7+commit.e28d00a7",
  "0.8.6": "0.8.6+commit.11564f7e",
  "0.8.5": "0.8.5+commit.a4f2e591",
  "0.8.4": "0.8.4+commit.c7e474f2",
  "0.8.3": "0.8.3+commit.8d00100c",
  "0.8.2": "0.8.2+commit.661d1103",
  "0.8.1": "0.8.1+commit.df193b15",
  "0.8.0": "0.8.0+commit.c7dfd78e"
 } as const;

/**
 * Input schema for deploy contract action.
 */
export const DeployContractInput = z
  .object({
    solidityVersion: z.enum(Object.keys(SolidityVersions) as [string, ...string[]]).describe("The solidity compiler version"),
    solidityInputJson: z.string().describe("The input json for the solidity compiler"),
    contractName: z.string().describe("The name of the contract class to be deployed"),
    constructorArgs: z
      .record(z.string(), z.any())
      .describe("The constructor arguments for the contract")
      .optional(),
  })
  .strip()
  .describe("Instructions for deploying an arbitrary contract");

/**
 * Deploys an arbitrary contract.
 *
 * @param wallet - The wallet to deploy the contract from.
 * @param args - The input arguments for the action. The three required fields are solidityVersion, solidityInputJson, and contractName. The constructorArgs field is only required if the contract has a constructor.
 * @returns A message containing the deployed contract address and details.
 */
export async function deployContract(
  wallet: Wallet,
  args: z.infer<typeof DeployContractInput>,
): Promise<string> {
  try {
    const solidityVersion = SolidityVersions[args.solidityVersion];

    const contract = await wallet.deployContract({
      solidityVersion: solidityVersion,
      solidityInputJson: args.solidityInputJson,
      contractName: args.contractName,
      constructorArgs: args.constructorArgs ?? {},
    });

    const result = await contract.wait();

    return `Deployed contract ${args.contractName} at address ${result.getContractAddress()}. Transaction link: ${result
      .getTransaction()!
      .getTransactionLink()}`;
  } catch (error) {
    return `Error deploying contract: ${error}`;
  }
}

/**
 * Deploy contract action.
 */
export class DeployContractAction implements CdpAction<typeof DeployContractInput> {
  public name = "deploy_contract";
  public description = DEPLOY_CONTRACT_PROMPT;
  public argsSchema = DeployContractInput;
  public func = deployContract;
}

File: cdp-agentkit-core/python/cdp_agentkit_core/actions/deploy_contract.py
from collections.abc import Callable

from cdp import Wallet
from pydantic import BaseModel, Field
from typing import Any

from cdp_agentkit_core.actions import CdpAction


DEPLOY_CONTRACT_PROMPT = """
This tool will deploy an arbitrary smart contract. It takes 3 required arguments: solidity compiler version which is a string, solidity input json which is a string, 
contract name which is a string, and 1 optional argument: constructor arguments which is a map of string to any.
The solidity version must be 0.8.+, such as "0.8.28+commit.7893614a" which is the latest version. See https://binaries.soliditylang.org/bin/list.json for valid versions.
The input json must be a valid solidity input json. See https://docs.soliditylang.org/en/latest/using-the-compiler.html#input-description for more details. If you 
depend on a library, make sure to include the library source code as separate sources inline in the input json.
For the settings outputSelection in the input JSON, make sure to have a string array including the abi and evm.bytecode.
The contract name must be the name of the contract class to be deployed.
The constructor arguments, if passed in, must be a map of constructor arguments for the contract, where the key is the argument name and the value is the argument value.
uint, int, bytes, fixed bytes, string, address should be encoded as strings. Boolean values should be encoded as true or false.
For arrays and tuples, the values should be encoded depending on the underlying type contained in the array or tuple.
"""


class DeployContractInput(BaseModel):
    """Input argument schema for deploy token action."""

    solidity_version: str = Field(..., description='The solidity compiler version')
    solidity_input_json: str = Field(..., description='The input json for the solidity compiler')
    contract_name: str = Field(..., description='The name of the contract class to be deployed')
    constructor_args: dict[str, Any] = Field(
        ..., description='The constructor arguments for the contract'
    )


def deploy_token(wallet: Wallet, name: str, symbol: str, total_supply: str) -> str:
    """Deploy an ERC20 token smart contract.

    Args:
        wallet (wallet): The wallet to deploy the Token from.
        name (str): The name of the token (e.g., "My Token")
        symbol (str): The token symbol (e.g., "USDC", "MEME", "SYM")
        total_supply (str): The total supply of tokens to mint (e.g., "1000000")

    Returns:
        str: A message containing the deployed token contract address and details

    """
    try:
        token_contract = wallet.deploy_token(name=name, symbol=symbol, total_supply=total_supply)

        token_contract.wait()
    except Exception as e:
        return f"Error deploying token {e!s}"

    return f"Deployed ERC20 token contract {name} ({symbol}) with total supply of {total_supply} tokens at address {token_contract.contract_address}. Transaction link: {token_contract.transaction.transaction_link}"


class DeployTokenAction(CdpAction):
    """Deploy token action."""

    name: str = "deploy_token"
    description: str = DEPLOY_TOKEN_PROMPT
    args_schema: type[BaseModel] | None = DeployTokenInput
    func: Callable[..., str] = deploy_token
